app/models/project_step.rb

```diff
@@ -196,6 +206,67 @@ def week_of_scheduled_date
     end
   end
 
+  # todo: confirm if we want to hardcode a limit for end_date bound repeats. and if so, what should that be?
+  MAX_OCCURRENCES = 100
+
+  def duplicate_series(frequency, time_unit, month_repeat_on, num_of_occurrences, end_date)
+    if time_unit == 'days' || time_unit == 'weeks'
+      interval = frequency.send(time_unit)
+    end
+
+    results = []
+    remaining = num_of_occurrences || MAX_OCCURRENCES
+
+    allow_error = true
+    last_date = duplicate_step_base_date
+    while remaining > 0 do
```

Comment:
Ruby way to do this would be `(num_of_occurrences || MAX_OCCURRENCES).times do`. Then you can remove the decrement line at the bottom.